test/tools/NamedPipeConnection/src/code/NamedPipeConnection.cs

```diff
@@ -0,0 +1,433 @@
+            textReader = _clientPipe.TextReader;
+        }
+
+        /// <summary>
+        /// Stops a connection attempt, or closes the connection that has been established.
+        /// </summary>
+        public void StopConnect()
+        {
+            _clientPipe?.AbortConnect();
+            _clientPipe?.Close();
+            _clientPipe?.Dispose();
+        }
+
+        #endregion
+    }
+
+    #endregion
+
+    #region NamedPipeClientSessionTransportMgr
+
+    internal sealed class NamedPipeClientSessionTransportMgr : ClientSessionTransportManagerBase
+    {
+        #region Fields
+
+        private readonly NamedPipeInfo _connectionInfo;
+        private const string _threadName = "NamedPipeCustomTransport Reader Thread";
+
+        #endregion
+
+        #region Constructor
+
+        internal NamedPipeClientSessionTransportMgr(
+            NamedPipeInfo connectionInfo,
+            Guid runspaceId,
+            PSRemotingCryptoHelper cryptoHelper)
+            : base(runspaceId, cryptoHelper)
+        {
+            if (connectionInfo == null) { throw new PSArgumentException("connectionInfo"); }
+
+            _connectionInfo = connectionInfo;
+        }
+
+        #endregion
+
+        #region Overrides
+
+        /// <summary>
+        /// Create a named pipe connection to the target process and set up
+        /// transport reader/writer.
+        /// </summary>
+        public override void CreateAsync()
+        {
+            _connectionInfo.Connect(
+                out StreamWriter pipeTextWriter,
+                out StreamReader pipeTextReader);
+
+            // Create writer for named pipe.
+            _messageWriter = new OutOfProcessTextWriter(pipeTextWriter);
+
+            // Create reader thread for named pipe.
+            StartReaderThread(pipeTextReader);
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            base.Dispose(isDisposing);
+
+            if (isDisposing)
+            {
+                CloseConnection();
+            }
+        }
+
+        protected override void CleanupConnection()
+        {
+            CloseConnection();
+        }
+
+        #endregion
+    
+        #region Private Methods
+
+        private void CloseConnection()
+        {
+            _connectionInfo.StopConnect();
+        }
+
+        private void HandleSSHError(PSRemotingTransportException ex)
+        {
+            RaiseErrorHandler(
+                new TransportErrorOccuredEventArgs(
+                    ex,
+                    TransportMethodEnum.CloseShellOperationEx));
+
+            CloseConnection();
+        }
+
+        private void StartReaderThread(
+            StreamReader reader)
+        {
+            Thread readerThread = new Thread(ProcessReaderThread);
+            readerThread.Name = _threadName;
+            readerThread.IsBackground = true;
+            readerThread.Start(reader);
+        }
+
+        private void ProcessReaderThread(object state)
```

Comment:
No, I think each custom derivation should have its own implementation.  The reason is that we cannot predict how underlying connection streams will work.  For example, 'NamedPipes' streams will throw 'expected' errors during connection close, that aren't really errors to be reported.  I want custom implementations to have full capability to deal with whatever connection stream they are using.

Reply:
My main concern was not around using a thread to read but rather the use of separate methods for sending input data to the "reader" `HandleErrorDataReceived` and `HandleOutputDataReceived`.

Wouldn't it be better to have transports just do something like

```csharp
try
{
    StreamReader reader = state as StreamReader;

    while (true)
    {
        string data = reader.ReadLine();
        if (data == null)
        {
            break;
        }

        HandleDataReceived(data);
    }
}
catch (ObjectDisposedException)
{
}
catch (Exception e)
{
    RaiseErrorHandler(...);
}
```

The transport can still inspect the data if it wishes but the minimal amount of work that it needs to do is just send the data from the peer to be handled. The code in `HandleDataReceived` can then raise it's own exception based on whether the payload was invalid or for whatever other reason. The code could even do `HandleDataReceived("Custom error message");` and the base transport class can set that error in a common way rather than something transport specific.